system/Encryption/Encryption.php

```diff
@@ -0,0 +1,298 @@
+use CodeIgniter\Config\Services;
+
+/**
+ * CodeIgniter Encryption Manager
+ *
+ * Provides two-way keyed encryption via PHP's Sodium and/or OpenSSL extensions.
+ * This class determines the driver, cipher, and mode to use, and then
+ * initializes the appropriate encryption handler.
+ */
+class Encryption
+{
+
+	/**
+	 * The encrypter we create
+	 *
+	 * @var string
+	 */
+	protected $encrypter;
+
+	/**
+	 * Our remembered configuration
+	 */
+	protected $config = null;
+
+	/**
+	 * Our default configuration
+	 */
+	protected $default = [
+		'driver' => 'OpenSSL', // The PHP extension we plan to use
+		'key'    => '', // no starting key material
+	];
+	protected $driver, $key;
+
+	/**
+	 * HMAC digest to use
+	 */
+	protected $digest = 'SHA512';
+
+	/**
+	 * Map of drivers to handler classes, in preference order
+	 *
+	 * @var array
+	 */
+	protected $drivers = [
+		'OpenSSL',
+	];
+
+	// --------------------------------------------------------------------
+
+	/**
+	 * Class constructor
+	 *
+	 * @param  mixed $params Configuration parameters
+	 * @return void
+	 *
+	 * @throws \CodeIgniter\Encryption\Exceptions\EncryptionException
+	 */
+	public function __construct($params = null)
+	{
+		$this->config = array_merge($this->default, (array) new \Config\Encryption());
+
+		if (is_string($params))
+		{
+			$params = ['driver' => $params];
+		}
+
+		$params = $this->properParams($params);
+
+		// Check for an unknown driver
+		if (isset($this->drivers[$params['driver']]))
+		{
+			throw EncryptionException::forDriverNotAvailable($params['driver']);
+		}
+
+		// determine what is installed
+		$this->handlers = [
+			'OpenSSL' => extension_loaded('openssl'),
+		];
+
+		if (! in_array(true, $this->handlers))
+		{
+			throw EncryptionException::forNoHandlerAvailable();
+		}
+	}
+
+	/**
+	 * Initialize or re-initialize an encrypter
+	 *
+	 * @param  array $params Configuration parameters
+	 * @return \CodeIgniter\Encryption\EncrypterInterface
+	 *
+	 * @throws \CodeIgniter\Encryption\Exceptions\EncryptionException
+	 */
+	public function initialize(array $params = [])
+	{
+		$params = $this->properParams($params);
+
+		// Insist on a driver
+		if (! isset($params['driver']))
+		{
+			throw EncryptionException::forNoDriverRequested();
+		}
+
+		// Check for an unknown driver
+		if (! in_array($params['driver'], $this->drivers))
+		{
+			throw EncryptionException::forUnKnownHandler($params['driver']);
+		}
+
+		// Check for an unavailable driver
+		if (! $this->handlers[$params['driver']])
+		{
+			throw EncryptionException::forDriverNotAvailable($params['driver']);
+		}
+
+		// Derive a secret key for the encrypter
+		if (isset($params['key']))
+		{
+			$hmacKey          = strcmp(phpversion(), '7.1.2') >= 0 ? \hash_hkdf($this->digest, $params['key']) : self::hkdf($params['key'], $this->digest);
+			$params['secret'] = bin2hex($hmacKey);
+		}
+
+		$handlerName     = 'CodeIgniter\\Encryption\\Handlers\\' . $this->driver . 'Handler';
```

Comment:
If we are moving drivers to a config class, we should define these classes there, also. Otherwise we get stuck in the situation we currently have with the database, where there's no easy way to extend it or write our own handlers for it.